projects/mpg123/read_fuzzer.c

```diff
@@ -0,0 +1,85 @@
+#include <stddef.h>
+#include <unistd.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <stdbool.h>
+#include <string.h>
+
+#include "mpg123.h"
+
+static char* fuzzer_get_tmpfile(const uint8_t* data, size_t size) {
+  char* filename_buffer = strdup("/tmp/generate_temporary_file.XXXXXX");
+  if (!filename_buffer) {
+    perror("Failed to allocate file name buffer.");
+    abort();
+  }
+  const int file_descriptor = mkstemp(filename_buffer);
+  if (file_descriptor < 0) {
+    perror("Failed to make temporary file.");
+    abort();
+  }
+  FILE* file = fdopen(file_descriptor, "wb");
+  if (!file) {
+    perror("Failed to open file descriptor.");
+    close(file_descriptor);
+    abort();
+  }
+  const size_t bytes_written = fwrite(data, sizeof(uint8_t), size, file);
```

Comment:
Forgot to mention, but it's really better not to do I/O when fuzzing. 
Much better if this can be done in memory.
What were the exec/s of this fuzzer when you ran it?

Reply:
Yeah I agree it isn't great. This is using a file because the file based API is simpler, but let me send out a non-file based version as well since I have it.

I'm not sure how to measure the executions per-second, but it generates 100 corpus elements in 4.7 seconds with a single job using libfuzzer.

Reply:
See `exec/s` here: https://llvm.org/docs/LibFuzzer.html#output

Reply:
Ah, thanks :)

On my local machine, the file-based fuzzer here runs at 4,500 exec/s, and oddly enough the non-file based fuzzer here https://github.com/google/oss-fuzz/pull/2037 runs at 4,000 exec/s.

Reply:
That seems fast enough.
I wonder why the non-file-based one is slower. I guess it is worth looking at the [stats](https://oss-fuzz.com/fuzzer-stats/by-fuzzer/fuzzer/libFuzzer/job/libfuzzer_asan_mpg123), once they start appearing to see if the non-file based target is faster over the long run (since speed can vary over time).